x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -2261,4 +2276,360 @@ private static AggregateMetricDoubleBlockBuilder.Metric getMetric(AggregateFunct
+            List<LogicalPlan> newChildren = new ArrayList<>(unionAll.children().size());
+            boolean outputChanged = false;
+            for (LogicalPlan child : unionAll.children()) {
+                List<Attribute> childOutput = child.output();
+                List<Alias> newAliases = new ArrayList<>();
+                List<Attribute> newChildOutput = new ArrayList<>(childOutput.size());
+                for (Attribute oldAttr : childOutput) {
+                    AbstractConvertFunction convertFunction = convertFunctions.get(oldAttr.name());
+                    // if the name of the attribute matches, and the data type is different, we need to add a conversion function
+                    if (convertFunction != null && oldAttr.dataType() != convertFunction.dataType()) {
+                        // create a new alias for the conversion function
+                        Alias newAlias = new Alias(
+                            oldAttr.source(),
+                            oldAttr.name(),
+                            convertFunction.replaceChildren(Collections.singletonList(oldAttr))
+                        );
+                        newAliases.add(newAlias);
+                        newChildOutput.add(newAlias.toAttribute());
+                        outputChanged = true;
+                    } else {
+                        newChildOutput.add(oldAttr);
+                    }
+                }
+                newChildren.add(maybePushDownConvertFunctionsToChild(child, newAliases, newChildOutput));
+            }
+
+            return outputChanged ? rebuildUnionAllOutput(unionAll, newChildren, convertFunctions, convertFunctionsToAttributes) : unionAll;
+        }
+
+        private Map<String, AbstractConvertFunction> collectConvertFunctions(UnionAll unionAll, LogicalPlan plan) {
+            Map<String, AbstractConvertFunction> convertFunctions = new HashMap<>();
+            plan.forEachDown(p -> p.forEachExpression(AbstractConvertFunction.class, f -> {
+                if (f.field() instanceof Attribute attr && unionAll.output().contains(attr)) {
+                    convertFunctions.putIfAbsent(attr.name(), f);
+                }
+            }));
+            return convertFunctions;
+        }
+
+        private LogicalPlan maybePushDownConvertFunctionsToChild(LogicalPlan child, List<Alias> aliases, List<Attribute> output) {
+            // Fork/UnionAll adds an EsqlProject on top of each child plan during resolveFork, check this pattern before pushing down
+            if (aliases.isEmpty() == false && child instanceof EsqlProject esqlProject) {
+                LogicalPlan childOfProject = esqlProject.child();
+                Eval eval = new Eval(childOfProject.source(), childOfProject, aliases);
+                return new EsqlProject(esqlProject.source(), eval, output);
+            }
+            return child;
+        }
+
+        private LogicalPlan rebuildUnionAllOutput(
```

Comment:
```suggestion
        private LogicalPlan rebuildUnionAll(
```

since it updates both children and output.
Alternatively, maybe keep the name, but move the construction in the calling method.